apps/prairielearn/src/pages/publicAssessmentQuestions/publicAssessmentQuestions.html.ts

```diff
@@ -0,0 +1,127 @@
+import { type Assessment, type Course } from '../../lib/db-types.js';
+import { type AssessmentQuestionRow } from '../../models/assessment-question.js';
+
+export function PublicAssessmentQuestions({
+  resLocals,
+  assessment,
+  course,
+  course_instance_id,
+  questions,
+}: {
+  resLocals: Record<string, any>;
+  assessment: Assessment;
+  course: Course;
+  course_instance_id: string;
+  questions: AssessmentQuestionRow[];
+}) {
+  return html`
+    <!doctype html>
+    <html lang="en">
+      <head>
+        ${HeadContents({ resLocals })}
+      </head>
+      <body>
+        ${Navbar({ resLocals })}
+        <main id="content" class="container">
+          ${course.sharing_name
+            ? html`
+                <div class="card mb-4">
+                  <div class="card-header bg-primary text-white d-flex align-items-center">
+                    ${assessment.title} ${assessment.number}: Questions
+                  </div>
+                  ${AssessmentQuestionsTable({
+                    questions,
+                    urlPrefix: resLocals.urlPrefix,
+                    course_id: course.id,
+                    course_instance_id,
+                    course_sharing_name: course.sharing_name,
+                  })}
+                </div>
+              `
+            : html`<p>
+                This course doesn't have a sharing name. If you are an Owner of this course, please
+                choose a sharing name on the
+                <a
+                  href="${resLocals.plainUrlPrefix}/course/${resLocals.course
+                    .id}/course_admin/sharing"
+                  >course sharing settings page</a
+                >.
+              </p>`}
+        </main>
+      </body>
+    </html>
+  `.toString();
+}
+
+function AssessmentQuestionsTable({
+  questions,
+  urlPrefix,
+  course_id,
+  course_instance_id,
+  course_sharing_name,
+}: {
+  questions: AssessmentQuestionRow[];
+  urlPrefix: string;
+  course_id: string;
+  course_instance_id: string;
+  course_sharing_name: string;
+}) {
+  const nTableCols = 4;
+  return html`
+    <div class="table-responsive">
+      <table class="table table-sm table-hover">
+        <thead>
+          <tr>
+            <th><span class="visually-hidden">Name</span></th>
+            <th>QID</th>
+            <th>Topic</th>
+            <th>Tags</th>
+            <th>Other Assessments</th>
+          </tr>
+        </thead>
+        <tbody>
+          ${questions.map((question) => {
+            return html`
+              ${AssessmentQuestionHeaders(question, nTableCols)}
+              <tr>
+                <td>
+                  <a
+                    href="${urlPrefix}/public/course/${course_id}/question/${question.question_id}/preview"
+                  >
+                    ${AssessmentQuestionNumber(question)}${question.title}
+                  </a>
+                </td>
+                <td>${`@${course_sharing_name}/${question.display_name}`}</td>
```

Comment:
I have some reservations about this. Namely: what happens when a shared assessment itself uses a shared question from another course? We have the following in `selectAssessmentQuestions`:

```sql
  CASE
    WHEN q.course_id = $course_id THEN qid
    ELSE '@' || c.sharing_name || '/' || q.qid
  END AS display_name
```

It seems to me like we would end up with something like `@foo/@bar/baz` in that case.

Reply:
oh wow, I totally forgot to think about this case. 

Can we just make it so that if the display_name already starts with a '@', then we don't but the course sharing name in front, but if it doesn't start with a '@' then we add it? 

The case when the assessment is using a shared question should be exactly the cases when the first character of display_name is '@'.

Reply:
the other alternative would be to pass another parameter into the `selectAssessmentQuestions`, denoting whether or not it is for the public assessment page, and then do that login in the SQL. Something like:

```sql
 CASE
    WHEN q.course_id = $course_id AND NOT $public THEN qid
    ELSE '@' || c.sharing_name || '/' || q.qid
  END AS display_name
```

do you have a preference?

Reply:
done in 2fdccfe57

Reply:
Unresolving, as I have a strong preference for moving this kind of logic up the stack into TypeScript when we actually go to render things.

Reply:
To move all the logic into the typescript, we would need to join the `course.sharing_name` onto every assessment question. Is that the way you want to do it?

Reply:
done in 2f7887180

Reply:
fixed bug in f636945fc5ff40f50f0149e35eae41388ae7f3e1 by using `idsEqual`